enhancements/olm/report-sdk-operators.md

```diff
@@ -0,0 +1,103 @@
+
+**Note**: 
+Stamps refer to the labels added as metadata in operator bundles sent to OLM. OLM would interpret these stamps and add labels to their existing metrics. 
+
+## Proposal
+
+The [proposal][sdk_metrics_proposal] lists the bundle resources in which stamps would be added indicating the use of SDK to build an operator bundle image. 
+
+The proposed stamps have the following format in `bundle.dockerfile` and `annotations.yaml`:
+1. operators.operatorframework.io.metrics.mediatype.v1 = “metrics+v1”
+2. operators.operatorframework.io.metrics.builder = “operator-sdk-sdkversion”
+3. operators.operatorframework.io.metrics.project_layout = “sdk_project_layout/layout_version” 
+
+In CSV and CRDs, they would have the following annotations in Object Metadata:
+1. operators.operatorframework.io/builder = “operator-sdk-sdkversion”
+2. operators.operatorframework.io/project_layout = “sdk_project_layout/layout_version”
+
+For example, in case of memcached-operator having Kubebuilder layout, the labels can be written as:
+
+```
+LABEL operators.operatorframework.io.metrics.mediatype.v1 = “metrics+v1”
+LABEL operators.operatorframework.io.metrics.builder = “operator-sdk-v0.17.0”
+LABEL operators.operatorframework.io.metrics.project_layout = “go.kubebuilder.io/v2.0.0”
+```
+
+The value of `project_layout` will represent the layout structure and version of SDK project for operators built using SDK after Kubebuilder integration. In case of operators using legacy SDK project structure, the value would represent the kind of operator (ie. `go`, `ansible` or `helm`)
+
+At present, some of the OLM metrics track the lifecycle of CSV present in operator bundle and report its status at various stages. These stamps can be parsed by OLM from bundle resources (CSV, CRDs, `annotations.yaml` or `bundle.Dockerfile`) and added to their metrics.
+
+For example, one such metric is `csvSucceeded`. Adding of `BUILDER` and `PROJECT_LAYOUT` label to this metric will enable us to track the version and layout of operator SDK which was used to build the bundle image.
+
+The metric could be modified to:
+
+```Go
+
+	// Prometheus labels added to metrics
+	const (
+		BUILDER = "builder"
+		PROJECT_LAYOUT = "project_layout"
+	)
+
+ 	csvSucceeded = prometheus.NewGaugeVec(
+ 		prometheus.GaugeOpts{
+ 			Name: "csv_succeeded",
+ 			Help: "Successful CSV install",
+ 		},
+ 		[]string{NAMESPACE_LABEL, NAME_LABEL, VERSION_LABEL, BUILDER, PROJECT_LAYOUT},
+    )
+```
+
+Here, the value of `BUILDER` would be taken from `operators.operatorframework.io/builder` and `PROJECT_LAYOUT` would represent the value of `operators.operatorframework.io/project_layout` from CSV annotations.
+
+An example prometheus metric value would be:
+
+```
+csvSucceeded{namespace="openshift-operator-lifecycle-manager", name="test-operator", version="x", builder="operator-sdk-v0.17.0", project_layout="go.kubebuilder.io/v2.0.0"}
```

Comment:
csvSucceeded is not the best metric name, I would change this to be something like `olm_csv_success_total`. I assume this will be incremented on each success?

See best practices -> https://prometheus.io/docs/practices/naming/#metric-names

Reply:
I agree. But these metric names are currently being used by OLM. (https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/metrics/metrics.go)

Reply:
IMO a metric name change is high cost and low value - queries that span versions of ocp become more painful.

but we should keep the naming conventions in mind for new metrics, and over time I imagine these poorly named ones will be obsoleted.